x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobResultsPersister.java

```diff
@@ -148,7 +151,7 @@ public Builder persistTimingStats(TimingStats timingStats) {
         public Builder persistRecords(List<AnomalyRecord> records) {
             for (AnomalyRecord record : records) {
                 logger.trace("[{}] ES BULK ACTION: index record to index [{}] with ID [{}]", jobId, indexName, record.getId());
-                indexResult(record.getId(), record, "record");
+                indexResult(record.getId(), record, FOR_INTERNAL_STORAGE, "record");
```

Comment:
This change hasn't had any effect as far as I can see, as the `toXContent` methods of `AnomalyRecord` and `AnomalyCause` don't check it.

This makes sense because so many clients (including Kibana) read anomalies direct out of our results index rather than calling the get_records endpoint that not writing the new fields to internal storage would make the feature pretty useless.

This does highlight one thing that is inconsistent between this implementation and everything else we've done with ML results in the index though.  We decided early on that the documents in the results index would have the same structure as the documents produced by the C++ process.  This change is moving away from that principle.

(The other case of results documents being augmented is adding Grok patterns to category definitions.  But that is not done in the results index; it's done in the transport action and only affects results obtained via the endpoint, not direct from the index.)

For jobs that run both before and after the upgrade to whatever version contains this functionality it will result in an inconsistency: results that are rewritten by renormalization will acquire the new fields even if they were generated by the old version of the code, whereas results that aren't rewritten by renormalization won't.

To make this consistent with what we've done for other result types the approach would be:

1. Generate the new fields in the C++
2. Just round-trip them through the X-Pack core Java classes, don't generate them there

Then records would have these fields if and only if they were generated by the version this change is released in.  The HLRC would stay as it is now.

On the other hand, that would cause the need to store two strings containing duplicate data in each `AnomalyRecord` and `AnomalyCause` object in memory.  The C++ output would include the strings and the Java class would have to store them, unaware that they contain the same information as existing fields but in a different format.  But then we don't store many of these objects in memory for a long time, so maybe that doesn't matter.

I think we should ask for other opinions about the pros and cons of letting what's stored in the index deviate from what's output by the C++ process.  It's not a big problem with this single change but at Prelert we let the two JSON structures diverge in many ways and then it became hard to remember what all the differences were (as people rarely view the JSON coming out of the C++) which led to subtle bugs after a couple of years of divergence.

Reply:
I also favour keeping objects the same between the 2 sides. It's a slippery slope if we allow divergence.

Reply:
The issue with inconsistent results is a painful one. Having it like it is now could also indicate to the user that fields support geo mapped values when they weren't actually indexed that way. This may cause them to think they are missing data somewhere.

Reply:
OK, so to have these fields created by the C++ they'd need adding to the `api::CJsonOutputWriter` class [here](https://github.com/elastic/ml-cpp/blob/master/lib/api/CJsonOutputWriter.cc#L555) (for non-population detectors) and [here](https://github.com/elastic/ml-cpp/blob/master/lib/api/CJsonOutputWriter.cc#L655) (for population detectors).  Would you like to make that change @benwtrent or would you be happier sticking to Java?

Either way, this Java PR would need to be merged first, and should be changed so that it doesn't create the values in Java code but instead stores them as `String`s in the results objects _only_ if they were present in the input JSON (which could come either from the C++ process or from reading docs out of the results index).